design doc 3-1-6 base station added (#44)
Capstone-Skynet/Capstone-Skynet.github.io: docs/design/3_1_6_base_station.tex

\textbf{Display the Result}

The main task of the base station is to verify the ML results. Different machine learning applications could have results in different forms. For object detection, the ML results will be displayed as a bounding box around the detected object in the frame. The ML result that the base station receives consists of two parts: coordinates of the upper-left vertex of the bounding box in the frame and values of height and width of the bounding box. The operations that the base station performs after receiving the data are:
\begin{enumerate}
    \item Calculates which pixels in the current frame form the desired bounding box and thus need to be re-programmed to a dedicated colour.
    \item Modify these pixels in the frame.
    \item Display the frame.
\end{enumerate}

> The program that is running on the base station is stored on the server - Raspberry Pi. When the base station connects to the server it automatically downloads the program and runs it. (\textbf{F.BS.1})

Reviewer: ```suggestion
The base station retrieves this information by querying a TCP/IP server hosted by the computing platform (Raspberry Pi) and displaying the results in a web-application (\textbf{F.BS.1}).
```

> The base station in the system is a mobile device that connects to the on-board computing platform, it is used for:

Reviewer: ```suggestion
The base station is a mobile device that connects to the computing platform on the multirotor. In particular, its applications are to:
```

\begin{itemize}

>     \item monitoring the status of the system.

Reviewer: ```suggestion
    \item monitor the status of the system,
```

>     \item verifying the ML result.

Reviewer: ```suggestion
    \item display the ML results in real-time (by overlaying the data over live video),
```

>     \item storing the log data and video for further analysis.

Reviewer: ```suggestion
    \item storing the log data and video for further analysis, and
```

>     \item control the computing platform, i.e. start object detection.

Reviewer: ```suggestion
    \item controlling the computing platform (ex. starting object detection).
```

\end{itemize}

\textbf{Options for the Base Station}

> First option is laptop. Laptop is a portable personal computer which has the following features:

Reviewer 2: ```suggestion
First option is a laptop. A laptop is a portable personal computer which has the following features:
```

> First option is a laptop. A laptop is a portable personal computer which has the following features:

Reviewer: I don't think this list is necessary - if the reader has gotten this far I think they know what a laptop is :P

Perhaps just discuss why a laptop would be advantageous over a mobile device in a short section.

\begin{itemize}
    \item It has a screen.
    \item It has a Wi-Fi transceiver. (\textbf{F.CM.3})
    \item It has a flash storage of at least 128GB.
    \item It is running Windows or macOS or Linux operating system, all of them are capable of running most video processing programs.
    \item It has an adequate processor for video processing.
    \item It is portable and has at least 3-hour battery life. (\textbf{F.BS.3})

> \end{itemize}

Reviewer 2: Capitalization consistency on \item

> The second option is tablet. Nowadays tablet is very similar to laptop in many ways. It has all the features listed above except it is running Android or iOS operating system. These mobile operating systems do not have support for running video processing programs. So it can not be used to have further analysis on the video data.

Reviewer 2: ```suggestion
The second option is a tablet. Nowadays tablets are very similar to laptops in many ways. It has all the features listed above except it is running Android or iOS operating system. These mobile operating systems do not have support for running video processing programs. So it can not be used to have further analysis on the video data.
```

> The second option is a tablet. Nowadays tablets are very similar to laptops in many ways. It has all the features listed above except it is running Android or iOS operating system. These mobile operating systems do not have support for running video processing programs. So it can not be used to have further analysis on the video data.

Reviewer: > These mobile operating systems do not have support for running video processing programs

This is not _necessarily_ true

\textbf{Base Station Setup}

> We choose laptop to be the primary base station. The primary base station is capable of doing all tasks of the base station. There could also have multiple secondary base stations. They can be either laptop or tablet. These secondary base stations may only be used to monitoring the status of the system and send control commands to the computing platform. Client can choose to use either of these base stations while testing but any further analysis on the data need to be done on the primary base station.

Reviewer 2: ```suggestion
We chose the laptop as the primary base station. The primary base station is capable of doing all tasks of the base station. There could also be multiple secondary base stations. They can be either laptop or tablet. These secondary base stations may only be used to monitoring the status of the system and send control commands to the computing platform. The client can choose to use either of these base stations while testing but any further analysis on the data need to be done on the primary base station.
```

> We chose the laptop as the primary base station. The primary base station is capable of doing all tasks of the base station. There could also be multiple secondary base stations. They can be either laptop or tablet. These secondary base stations may only be used to monitoring the status of the system and send control commands to the computing platform. The client can choose to use either of these base stations while testing but any further analysis on the data need to be done on the primary base station.

Reviewer: How would we synchronize commands across base-stations? 

Perhaps instead say that, if time permits, we'll extend limited functionality to viewing results on tablets.

\textbf{Display the Result}

> The main task of the base station is to verify the ML result. On the base station, the ML result will be displayed as a bounding box around the detected object in the frame. The ML result that the base station receives consists of two parts: coordinates of the upper-left vertex of the bounding box in the frame and values of height and width of the bounding box. The operations that the base station performs after receiving the data are:

Reviewer: You should preface this by saying that this is assuming a particular machine learning implementation. Our final implementation may not actually use bounding boxes.

\begin{enumerate}
    \item Calculates which pixels in the current frame form the desired bounding box and thus need to be re-programmed to a dedicated colour.
    \item Modify these pixels in the frame.
    \item Display the frame.
\end{enumerate}

> The program that is running on the base station is stored on the server - Raspberry Pi. When the base station connects to the server it automatically downloads the program and runs it. (\textbf{F.BS.1})

Reviewer: Is this true?

If the Raspberry Pi is the server and we're implementing a web-app, we should describe it that way (we're not "downloading a program").

Author: I am not sure about this. When we were testing the web-app that Muchen wrote and video streaming, all the code is on the Raspberry Pi. On the client side we just opened a web browser and connect to the server. That's why I wrote it's automatically downloaded from the server.

Reviewer 3: I think "serve" is a better word. There exists client-side HTML and JS code that's served to the client device.

\textbf{System Fault}

> Another role of the base station is to alert client about any system faults occurred. All the system faults, either detected by the base station itself or received fault message from the Raspberry Pi, is displayed on the base station's screen. In case of a fault that has a automatic fixing protocol, e.g. reconnect to the same Wi-Fi if lost connection, the base station will perform the corresponding operations and the warning message will be updated accordingly.

Reviewer: ```suggestion
Another role of the base station is to alert the end-user to any system warnings or errors. All such faults, either detected by the the computing platform or the base station itself, is displayed on the base station's screen. All faults will also be logged locally on the device to facilitate further debugging if necessary.
```
